## Abro el csv

In [35]:
import pandas as pd

# abrimos el archivo usando una función específica de pandas
dataset_original = pd.read_csv("covers.csv")
#print(raw_dataset.info())

## Comenzamos viendo los datos que tiene el dataset

In [36]:
#Creo una copia para no modificar el dataset original
DataSetCopia = dataset_original.copy()
DataSetCopia.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 998 entries, 0 to 997
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Track             998 non-null    object 
 1   Artist            998 non-null    object 
 2   Duration          998 non-null    object 
 3   Time_Signature    998 non-null    int64  
 4   Danceability      998 non-null    float64
 5   Energy            998 non-null    float64
 6   Key               998 non-null    int64  
 7   Loudness          998 non-null    float64
 8   Mode              998 non-null    int64  
 9   Speechiness       998 non-null    float64
 10  Acousticness      998 non-null    float64
 11  Instrumentalness  998 non-null    float64
 12  Liveness          998 non-null    float64
 13  Valence           998 non-null    float64
 14  Tempo             998 non-null    float64
 15  Popularity        998 non-null    int64  
 16  Year              998 non-null    int64  
dt

## Vemos los primeros elementos para comprender la estructura de los datos

In [37]:
DataSetCopia.head()

Track              Artist Duration  Time_Signature  \
0                   Babe                Styx     3:38               4   
1               The Rose        Bette Midler     4:04               4   
2                   Cars          Gary Numan     4:08               4   
3                  Magic  Olivia Newton-John     2:17               4   
4  We Don’t Talk Anymore       Cliff Richard     3:37               4   

   Danceability  Energy  Key  Loudness  Mode  Speechiness  Acousticness  \
0         0.700   0.582   11    -5.960     0       0.0356       0.05020   
1         0.264   0.640    8    -6.221     1       0.0442       0.03930   
2         0.338   0.562    9    -7.181     1       0.0290       0.03900   
3         0.911   0.689    1    -6.176     1       0.2650       0.00119   
4         0.728   0.563    1    -8.053     0       0.1340       0.62100   

   Instrumentalness  Liveness  Valence    Tempo  Popularity  Year  
0          0.000000    0.0881    0.785  116.712          96  1980  
1          0.000002    0.1510    0.190   84.828          92  1980  
2          0.000000    0.1070    0.259  149.907          82  1980  
3          0.000000    0.0704    0.546  140.034          80  1980  
4          0.000000    0.1790    0.352  100.017          80  1980

In [38]:
DataSetCopia.describe()

Time_Signature  Danceability      Energy         Key    Loudness  \
count      998.000000    998.000000  998.000000  998.000000  998.000000   
mean         3.966934      0.626469    0.633362    5.225451   -8.886123   
std          0.228178      0.151517    0.203787    3.657376    3.829896   
min          1.000000      0.174000    0.018300    0.000000  -28.980000   
25%          4.000000      0.534000    0.489000    2.000000  -11.256000   
50%          4.000000      0.633000    0.651500    5.000000   -8.270000   
75%          4.000000      0.735000    0.797000    9.000000   -6.045000   
max          5.000000      0.988000    0.994000   11.000000   -1.496000   

             Mode  Speechiness  Acousticness  Instrumentalness    Liveness  \
count  998.000000    998.00000    998.000000        998.000000  998.000000   
mean     0.689379      0.05763      0.244627          0.042554    0.178618   
std      0.462980      0.05574      0.248634          0.157057    0.162648   
min      0.000000      0.02270      0.000003          0.000000    0.022300   
25%      0.000000      0.03170      0.043350          0.000000    0.083950   
50%      1.000000      0.03935      0.155000          0.000022    0.113500   
75%      1.000000      0.05645      0.387250          0.001448    0.225750   
max      1.000000      0.52400      0.996000          0.974000    0.981000   

          Valence       Tempo  Popularity         Year  
count  998.000000  998.000000  998.000000   998.000000  
mean     0.603099  120.904174   57.743487  1984.498998  
std      0.257874   26.254971   17.397193     2.873372  
min      0.028700   61.530000    0.000000  1980.000000  
25%      0.388000  102.303000   49.000000  1982.000000  
50%      0.644000  119.969000   60.000000  1984.500000  
75%      0.824750  135.000750   70.000000  1987.000000  
max      0.984000  208.571000   96.000000  1989.000000

Daceability: podemos ver que la media y la mediana(50%) son valores muy cercanos y el maximo esta mas cerca de la mediana que el minimo

Energy: el minimo esta muy por debajo de la media, posible outlier

Key:pareciera una distribucion bastante normal

Loudness: Luego de investigar la variable, vimos que los valores negativos son comunes en la representacion de decibeles para las canciones

Mode: por lo que se ve en la media(0.689) podemos ver que la mayoria de la canciones esta en modo mayor

Speechiness: podemos ver un posible outlier en el valor maximo, esto podría ser una canción muy hablada, o bien un error de tipeo y que se corrió una coma, lo analizaremos mas tarde.

Acousticness: el valor minimo podría ser un outlier, y el maximo también aunque en menor medida

Instrumentalness: el valor maximo posible outlier y muchos valores en 0 (podiran ser canciones de solo voces)

Liveness: son valores entre 0 y 1 por ser probabilidades, la media de 0.178618 indica una muestra sesgada a derecha

Valence: valores entre 0 y 1, poqueño sesgo a izquierda

Tempo, Popularity y Year: distribucion bastante normal

# Limpieza y preprocesamiento de datos

## Busqueda de valores nulos

In [39]:
#Es pra buscar nulos
print(DataSetCopia.isna().sum())


Track               0
Artist              0
Duration            0
Time_Signature      0
Danceability        0
Energy              0
Key                 0
Loudness            0
Mode                0
Speechiness         0
Acousticness        0
Instrumentalness    0
Liveness            0
Valence             0
Tempo               0
Popularity          0
Year                0
dtype: int64


Por el momento pareciera que no hay elementos nulos porque no tenemos ningun valor con un nan

Posteriormente analizamos la existencia de valores nulos para las variables numéricas, no encontramos ningun potencial nulo pero encontramos dos columnas con valores a analizar.

In [40]:
DataSetCopia["Popularity"].value_counts().sort_index()

Popularity
0     12
1      3
3      2
4      2
5      1
      ..
90     1
91     2
92     2
93     1
96     1
Name: count, Length: 90, dtype: int64

Tenemos 12 canciones que tiene popularidad 0, esto podría representar varias cosas, pero al ser 12 de 998 filas que tenemos y al no saber como se mide esta variable, asumimos que el 0 es un valor posible para esta columna; Podría representar el grado de popularidad más bajo.

In [41]:
DataSetCopia["Instrumentalness"].value_counts().sort_index()

Instrumentalness
0.000000    307
0.000001      1
0.000001      2
0.000001      1
0.000001      1
           ... 
0.931000      1
0.939000      1
0.953000      1
0.964000      1
0.974000      1
Name: count, Length: 605, dtype: int64

Con la Columna 'instrumentalness' encontramos muchas filas que tienen un 0, sin embargo consideramos que no se trata de valores nulos, si no que puede tratarse de canciones cantadas acapella

Ahora vamos a buscar elementos que potencialmente representan un nulo en las columnas track y artist

In [42]:
DataSetCopia["Track"].value_counts()
#for idx, val in DataSetCopia["Track"].value_counts().items():
  #print(idx, val)
  

Track
Gloria                                   2
Physical                                 2
Don’t You Want Me                        2
Cherish                                  2
I Feel For You                           2
                                        ..
Always Something There To Remind Me      1
Overkill                                 1
Wanna Be Startin’ Somethin’              1
Baby, Come To Me                         1
Baby, I Love Your Way/Freebird Medley    1
Name: count, Length: 972, dtype: int64

In [43]:
for idx, val in DataSetCopia["Artist"].value_counts().items():
  print(idx, val)

Madonna 17
Daryl Hall & John Oates 13
Michael Jackson 13
Lionel Richie 12
Prince 11
Stevie Wonder 9
Bruce Springsteen 9
Elton John 9
Huey Lewis and the News 9
Billy Joel 9
Whitney Houston 8
Air Supply 8
Kool & the Gang 8
Duran Duran 8
Phil Collins 8
Kenny Loggins 7
Bon Jovi 7
The Pointer Sisters 7
Foreigner 7
Sheena Easton 7
Kenny Rogers 7
Billy Ocean 7
George Michael 6
Cyndi Lauper 6
Genesis 6
John Cougar Mellencamp 6
Janet Jackson 6
Chicago 6
Exposé 6
Olivia Newton-John 6
Bobby Brown 6
The Police 6
Culture Club 6
Rod Stewart 6
Diana Ross 6
Debbie Gibson 5
Richard Marx 5
Rick Springfield 5
Journey 5
Steve Winwood 5
The Cars 5
Bryan Adams 5
The Bangles 5
New Kids on the Block 5
Juice Newton 5
The Jets 5
Neil Diamond 5
Heart 5
Styx 5
Van Halen 5
Little River Band 5
Tina Turner 5
Dan Fogelberg 5
Laura Branigan 5
Donna Summer 5
INXS 5
Christopher Cross 5
Don Henley 4
REO Speedwagon 4
Gloria Estefan and Miami Sound Machine 4
Commodores 4
Taylor Dayne 4
Tears for Fears 4
Pat Benatar 4
Queen

Pudimos notar que tanto en la columna "Track" como en "Artist" no encontramos nulos ni valores que representen valores nulos.

### Creación una nueva columna a partir de otras

La columna "Artist" tiene colaboraciones entre artistas. Si bien puede ser de utilidad separar los artistas para poder realizar analisis para cada uno, la cantidad de valores únicos sería realmente grande, por lo que generar variables dummy por cada artista no es una opción.
Una idea que se nos ocurre, es separar la columna 'Artist' en varias columnas llamadas 'Artist_1', 'Artist_2', etc, para poder ingresar más de un artista para una única canción. O agregar una columna "Feat" para indicar que es una colaboración. Esta última opción es la que consideramos más valiosa.

In [44]:
columnaArtistas = DataSetCopia['Artist']
artistasUnicos = columnaArtistas.unique()
print(artistasUnicos)

['Styx' 'Bette Midler' 'Gary Numan' 'Olivia Newton-John' 'Cliff Richard'
 'Andy Gibb' 'Doobie Brothers' 'Michael Jackson' 'Fleetwood Mac'
 'Diana Ross' 'Rupert Holmes' 'Kenny Rogers' 'Billy Joel' 'Donna Summer'
 'Commodores' 'Queen' 'Dan Fogelberg' 'KC and the Sunshine Band'
 'Air Supply' 'Blondie' 'Anne Murray' 'Smokey Robinson' 'Kim Carnes'
 'Kenny Loggins' 'Barbra Streisand' 'Bob Seger and the Silver Bullet Band'
 'Christopher Cross' 'Ambrosia' 'Captain & Tennille' 'Eagles'
 'Linda Ronstadt' 'Jefferson Starship' 'The Manhattans'
 'Tom Petty and the Heartbreakers' 'George Benson' 'Irene Cara'
 'Paul McCartney' 'Lipps Inc' 'Kool & the Gang' 'Stevie Wonder'
 'The Brothers Johnson' 'Pete Townshend' 'Carole King' 'Robbie Dupree'
 'Little River Band' 'Benny Mardones' 'Eddie Rabbitt' 'Foreigner' 'Prince'
 'Johnny Lee' 'Paul Simon' 'Elton John' 'Dr. Hook' 'Teri DeSario & KC'
 'The Rolling Stones' 'the Pointer Sisters' 'Shalamar' 'The Spinners'
 'Olivia Newton-John / Electric Light Orchestra

In [45]:
#Con los datos de arriba pudimos ver los caracteres que se usan para separar a los artistas
DataSetCopia['Featuring'] = DataSetCopia['Artist'].apply(lambda x: (" and " in x) or (" & " in x) or ("/" in x) or (", " in x) or (" with " in x))

#DataSetCopia.info()


In [46]:
#Aca podemos ver un ejemplo de una colaboracion y como en la columna 'featuring' hay un true
fila = DataSetCopia.loc[DataSetCopia['Artist'] == 'KC and the Sunshine Band']
print(fila)

              Track                    Artist Duration  Time_Signature  \
18  Please Don’t Go  KC and the Sunshine Band     3:01               4   

    Danceability  Energy  Key  Loudness  Mode  Speechiness  Acousticness  \
18         0.683   0.554    6    -7.487     0         0.05         0.379   

    Instrumentalness  Liveness  Valence   Tempo  Popularity  Year  Featuring  
18           0.00011    0.0845    0.452  85.455          74  1980       True  


### Tratamiento de la columna Duration

En el caso de la columna "Duration" nos encontramos con valores del tipo String, que deberían ser tratados como valores Ordinales. Para esto, reemplazamos el tiempo que se encuentra representado por minutos:segundos, a segundos.

In [47]:
def duration_to_seconds(duration):
    minutes, seconds = map(int, duration.split(':'))
    return minutes * 60 + seconds

DataSetCopia['Duration_seconds'] = DataSetCopia['Duration'].apply(duration_to_seconds)

#Borrar duracion
DataSetCopia = DataSetCopia.drop('Duration', axis = 1)

DataSetCopia['Featuring'].value_counts()

Featuring
False    888
True     110
Name: count, dtype: int64

## Busqueda elementos repetidos

In [48]:
print(DataSetCopia.duplicated().sum())
duplicados = DataSetCopia[DataSetCopia.duplicated()]
duplicados

1


Track         Artist  Time_Signature  Danceability  Energy  Key  \
589  Sugar Walls  Sheena Easton               4         0.634   0.526    1   

     Loudness  Mode  Speechiness  Acousticness  Instrumentalness  Liveness  \
589   -10.281     0       0.0531         0.612            0.0102     0.127   

     Valence  Tempo  Popularity  Year  Featuring  Duration_seconds  
589    0.729  86.78          43  1985      False               123

Aca encontramos que hay una única fila repetida, la cual es la de la canción Sugar Walls. Por lo tanto una de estas filas tiene que ser eliminada ya que no nos brinda informacion necesaria.

In [49]:
DataSetCopia[DataSetCopia.duplicated(keep=False)]

# eliminamos los duplicados
DataSetCopia.drop_duplicates(keep="first", inplace=True)
# contamos la cantidad de duplicados
print(DataSetCopia.duplicated().sum())

0


Ahora vamos a ver las canciones que tengan el mismo nombre y artista, ya que podrian representar canciones que esten repetidas

In [50]:
duplicados = DataSetCopia[DataSetCopia.duplicated(subset=['Track', 'Artist'], keep=False)]
duplicados.sort_values(by= ['Danceability'])

Track              Artist  Time_Signature  \
223  Waiting For A Girl Like You           Foreigner               4   
125  Waiting For A Girl Like You           Foreigner               4   
12                          Lady        Kenny Rogers               4   
114                         Lady        Kenny Rogers               4   
208                     Physical  Olivia Newton-John               4   
113                     Physical  Olivia Newton-John               4   
451               I Feel For You          Chaka Khan               4   
553               I Feel For You          Chaka Khan               4   
304                  Uptown Girl          Billy Joel               4   
401                  Uptown Girl          Billy Joel               4   
453   All Night Long (ALL Night)       Lionel Richie               4   
352   All Night Long (ALL Night)       Lionel Richie               4   
227                       Gloria      Laura Branigan               4   
332                       Gloria      Laura Branigan               4   
50      Master Blaster (JAMMIN’)       Stevie Wonder               4   
145     Master Blaster (JAMMIN’)       Stevie Wonder               4   
121   Another One Bites The Dust               Queen               4   
16    Another One Bites The Dust               Queen               4   

     Danceability  Energy  Key  Loudness  Mode  Speechiness  Acousticness  \
223         0.330   0.469   10    -9.286     0       0.0335        0.3100   
125         0.330   0.469   10    -9.286     0       0.0335        0.3100   
12          0.539   0.160    3   -13.285     1       0.0329        0.8810   
114         0.539   0.160    3   -13.285     1       0.0329        0.8810   
208         0.647   0.844    0    -3.756     1       0.0457        0.0137   
113         0.647   0.844    0    -3.756     1       0.0457        0.0137   
451         0.698   0.912   11    -8.993     1       0.0675        0.0145   
553         0.698   0.912   11    -8.993     1       0.0675        0.0145   
304         0.701   0.944    4    -2.986     1       0.0455        0.0755   
401         0.701   0.944    4    -2.986     1       0.0455        0.0755   
453         0.744   0.583    1   -12.595     1       0.0429        0.1530   
352         0.744   0.583    1   -12.595     1       0.0429        0.1530   
227         0.776   0.743    9    -6.807     1       0.0289        0.1070   
332         0.776   0.743    9    -6.807     1       0.0289        0.1070   
50          0.874   0.419    5   -14.905     0       0.0563        0.0467   
145         0.874   0.419    5   -14.905     0       0.0563        0.0467   
121         0.932   0.528    5    -6.472     0       0.1610        0.1150   
16          0.932   0.528    5    -6.472     0       0.1610        0.1150   

     Instrumentalness  Liveness  Valence    Tempo  Popularity  Year  \
223          0.000002    0.0402    0.433  201.327          71  1982   
125          0.000002    0.0402    0.433  201.327          71  1981   
12           0.000268    0.1050    0.322   81.446          76  1980   
114          0.000268    0.1050    0.322   81.446          76  1981   
208          0.000658    0.1020    0.746  146.967          78  1982   
113          0.000658    0.1020    0.746  146.967          78  1981   
451          0.022500    0.2790    0.797  124.917          59  1984   
553          0.022500    0.2790    0.797  124.917          59  1985   
304          0.000000    0.6010    0.792  128.993          85  1983   
401          0.000000    0.6010    0.792  128.993          85  1984   
453          0.000078    0.0248    0.852  108.792          59  1984   
352          0.000078    0.0248    0.852  108.792          59  1983   
227          0.293000    0.1490    0.840  131.343          68  1982   
332          0.293000    0.1490    0.840  131.343          68  1983   
50           0.000510    0.1940    0.971  131.159          60  1980   
145          0.000510    0.1940    0.971  131.159         

En un principio consideramos que las canciones son la misma pero en una especie de remake, pero esta opcion no parece convincente porque todas las demas columnas son iguales. Podríamos suponer que la misma canción se publico dos veces en distintos años, lo cual no representa una informacion importante y podríamos borrar una de estas filas duplicadas

Sin embargo, la cantidad de filas que aparecen duplicadas son pocas en relación a la cantidad total del data set por lo que decidimos no eliminarlas ya que no afectarán en exceso el análisis.

## Comienzo del analisis de los datos

In [51]:
import ydata_profiling as ydp

# generamos el reporte
report = ydp.ProfileReport(DataSetCopia)
# lo mostramos interactivo en nuestra notebook
report.to_notebook_iframe()
# pasamos el reporte a un archivo html
report.to_file(output_file="Reporte_Spotify.html")


AttributeError: module 'pkgutil' has no attribute 'ImpImporter'